include/lapack/orghr.hpp

```diff
@@ -0,0 +1,98 @@
+/// @file orghr.hpp
+/// @author Thijs Steel, KU Leuven, Belgium
+/// Adapted from @see https://github.com/Reference-LAPACK/lapack/tree/master/SRC/dorghr.f
+//
+// This file is part of <T>LAPACK.
+// <T>LAPACK is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#ifndef __ORGHR_HH__
+#define __ORGHR_HH__
+
+#include "lapack/utils.hpp"
+#include "lapack/types.hpp"
+#include "lapack/larf.hpp"
+#include "lapack/org2r.hpp"
+
+namespace lapack {
+
+/** Generates a m-by-n matrix Q with orthogonal columns.
+ * 
+ * @param[in] ilo integer
+ * @param[in] ihi integer
+ *      ilo and ihi must have the same values as in the
+ *      previous call to gehrd. Q is equal to the unit
+ *      matrix except in the submatrix Q(ilo+1:ihi,ilo+1:ihi).
+ *      0 <= ilo <= ihi <= max(1,n).
+ * @param work Vector of size n-1.
+ * 
+ * @ingroup gehrd
+ */
+template<
+    class matrix_t, class vector_t, class work_t >
+int orghr(
+    size_type< matrix_t > ilo,
+    size_type< matrix_t > ihi,
+    matrix_t& A,
+    vector_t& tau,
+    work_t& work )
+{
+    using blas::scal;
+    using T      = type_t< matrix_t >;
+    using idx_t  = size_type< matrix_t >;
+    using pair  = std::pair<idx_t,idx_t>;
+    
+    // constants
+    const T zero( 0.0 );
+    const T one ( 1.0 );
+    const idx_t m = nrows(A);
+    const idx_t n = ncols(A);
+
+    // check arguments
+    lapack_error_if( (idx_t) size(tau)  < std::min<idx_t>( m, n ), -2 );
+    lapack_error_if( (idx_t) size(work) < n-1, -3 );
+
+
+    // Shift the vectors which define the elementary reflectors one
+    // column to the right, and set the first ilo and the last n-ihi
+    // rows and columns to those of the unit matrix
+
+    // This is currently optimised for column matrices, it may be interesting
+    // to also write these loops for row matrices
```

Comment:
I want to comment on that. No action required. But if we want to specialize this routine for row-major matrices, we just need to replace

```c++
template<
    class matrix_t, class vector_t, class work_t >
int orghr( ... ) { ... }
```

by 

```c++
template<
    class matrix_t, class vector_t, class work_t,
    enable_if_t<
        ! is_same_v< layout_type<matrix_t>, RowMajor_t >
    , int > = 0
>
int orghr( ... ) { ... }
```

and add a new routine for row major:

```c++
template<
    class matrix_t, class vector_t, class work_t,
    enable_if_t<
        is_same_v< layout_type<matrix_t>, RowMajor_t >
    , int > = 0
>
int orghr( ... ) { ... }
```